### PR TITLE
sg: update binary correctly instead of renaming file

### PR DESCRIPTION
**Pull request by mrnugget**
Linux users on Slack reported that `sg update` produces this error:

    >  sg update
    error: rename /tmp/sg1226370089/sg /home/fkling/.local/bin/sg: invalid cross-device link

As Geoffrey pointed out:

> rename only works within the same filesystem. I think the implementation should just copy / delete the file instead.

And the [man page for rename](https://linux.die.net/man/2/rename) says:

> oldpath and newpath are not on the same mounted file system. (Linux permits a file system to be mounted at multiple points, but rename() does not work across different mount points, even if the same file system is mounted on both.)

So what this here does is to use code from `gitserver` that does an
in-place update of a file safele. I extracted the code into an
`internal/fileutil` package (couldn't think of a better name).

The code avoids the across-filesystem problem by creating a temp file
next to the destination file. And it also does more things, like making
sure the changes are synced to the filesystem.

## Test plan

In `./dev/sg`:

```
$ go build -o ~/bin/sg . && sg version && sg update && sg version
dev
✅ sg has been updated!
To see what's new, run 'sg version changelog'.
d8176d40f3ddf54d47317c18f9d1532293eb6c23
```

### CODE REVIEW

File: internal/fileutil/fileutil.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	if n, err := f.Write(content); err != nil {
+		f.Close()
+		return false, err
+	} else if n != len(content) {
+		f.Close()
+		return false, io.ErrShortWrite
+	}
+
+	// fsync to ensure the disk contents are written. This is important, since
+	// we are not guaranteed that os.Rename is recorded to disk after f's
```
Comment (jhchabran) [inline internal/fileutil/fileutil.go:40]: Isn't a word missing here? Or that's just me not being used to use "after" like this?

**Review by jhchabran — APPROVED** [review]
TIL rename and FS!

File: internal/fileutil/fileutil.go (hunk 1 above)
Comment (mrnugget) [reply internal/fileutil/fileutil.go:40]: I can't spot a missing word? But maybe I'm missing something? 😄  (Also: I just copied these functions 1:1 and only changed the first letter of their name to export them)

**Review by keegancsmith — COMMENTED** [review]
LGTM

**Pull request merged**